File: backend/schemas/bank_recon.py
"""Request/response schemas for bank reconciliation."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class BankReconProcessRequest(BaseModel):
    period: str = Field(..., description="Tax period, e.g. '2026-01'")
    session_id: UUID | None = Field(None, description="Link to existing recon session")
    amount_tolerance: float = Field(0.01, ge=0)
    date_tolerance_days: int = Field(3, ge=0, le=30)
    run_ai_analysis: bool = True


class SourceFileInfo(BaseModel):
    filename: str
    file_type: str
    format_detected: bool
    bank_name: str | None
    row_count: int


class MatchPair(BaseModel):
    match_group_id: str
    record_id: str
    bank_id: str
    record_amount: float
    bank_amount: float
    date_diff_days: int | None = None


class UnmatchedEntry(BaseModel):
    id: str
    amount: float
    date: str | None = None
    description: str | None = None


class MatchResultResponse(BaseModel):
    matched_pairs: list[MatchPair] = []
    unmatched_records: list[UnmatchedEntry] = []
    unmatched_bank: list[UnmatchedEntry] = []
    match_rate: float = 0.0
    total_records: int = 0
    total_bank_entries: int = 0


class AISuggestion(BaseModel):
    unmatched_entry_index: int
    suggested_record_id: str | None = None
    confidence: float
    explanation: str
    category: str
    status: str = "pending"  # pending | accepted | rejected


class AIExplanation(BaseModel):
    entry_index: int
    entry_type: str
    mismatch_type: str
    explanation: str
    recommended_action: str


class BankReconBatchResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    created_by: UUID
    session_id: UUID | None = None
    status: str
    source_files: list[SourceFileInfo] | None = None
    total_entries: int = 0
    parse_summary: dict | None = None
    match_result: MatchResultResponse | None = None
    ai_suggestions: list[AISuggestion] | None = None
    ai_explanations: list[AIExplanation] | None = None
    amount_tolerance: float = 0.01
    date_tolerance_days: int = 3
    period: str
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BankReconBatchListItem(BaseModel):
    id: UUID
    status: str
    period: str
    total_entries: int = 0
    source_files: list[SourceFileInfo] | None = None
    match_rate: float | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class SuggestionActionRequest(BaseModel):
    suggestion_index: int = Field(..., ge=0)

File: frontend/src/api/bankRecon.ts
import { client } from './client'

export interface SourceFileInfo {
  filename: string
  file_type: string
  format_detected: boolean
  bank_name: string | null
  row_count: number
}

export interface MatchPair {
  match_group_id: string
  record_id: string
  bank_id: string
  record_amount: number
  bank_amount: number
  date_diff_days: number | null
}

export interface UnmatchedEntry {
  id: string
  amount: number
  date: string | null
  description: string | null
}

export interface MatchResult {
  matched_pairs: MatchPair[]
  unmatched_records: UnmatchedEntry[]
  unmatched_bank: UnmatchedEntry[]
  match_rate: number
  total_records: number
  total_bank_entries: number
}

export interface AISuggestion {
  unmatched_entry_index: number
  suggested_record_id: string | null
  confidence: number
  explanation: string
  category: string
  status: string
}

export interface AIExplanation {
  entry_index: number
  entry_type: string
  mismatch_type: string
  explanation: string
  recommended_action: string
}

export interface BankReconBatch {
  id: string
  tenant_id: string
  created_by: string
  session_id: string | null
  status: string
  source_files: SourceFileInfo[] | null
  total_entries: number
  parse_summary: Record<string, unknown> | null
  match_result: MatchResult | null
  ai_suggestions: AISuggestion[] | null
  ai_explanations: AIExplanation[] | null
  amount_tolerance: number
  date_tolerance_days: number
  period: string
  error_message: string | null
  created_at: string
  updated_at: string
}

export interface BankReconBatchListItem {
  id: string
  status: string
  period: string
  total_entries: number
  source_files: SourceFileInfo[] | null
  match_rate: number | null
  created_at: string
}

export const bankReconApi = {
  async process(formData: FormData) {
    const { data } = await client.post('/bank-recon/process', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 120000,
    })
    return data
  },

  async listBatches(page = 1, limit = 20) {
    const { data } = await client.get('/bank-recon/batches', {
      params: { page, limit },
    })
    return data
  },

  async getBatch(batchId: string) {
    const { data } = await client.get(`/bank-recon/batches/${batchId}`)
    return data
  },

  async acceptSuggestion(batchId: string, suggestionIndex: number) {
    const { data } = await client.post(
      `/bank-recon/batches/${batchId}/accept-suggestion`,
      { suggestion_index: suggestionIndex }
    )
    return data
  },

  async rejectSuggestion(batchId: string, suggestionIndex: number) {
    const { data } = await client.post(
      `/bank-recon/batches/${batchId}/reject-suggestion`,
      { suggestion_index: suggestionIndex }
    )
    return data
  },

  async rerunAnalysis(batchId: string) {
    const { data } = await client.post(
      `/bank-recon/batches/${batchId}/rerun-analysis`
    )
    return data
  },
}
